src/sql/SqlService.ts

```diff
@@ -0,0 +1,290 @@
+    }
+
+    /**
+     * Handles SQL execute response.
+     * @param clientMessage The response message
+     * @param res SQL result for this response
+     */
+    handleExecuteResponse(clientMessage: ClientMessage, res: SqlResultImpl): void {
+        const response = SqlExecuteCodec.decodeResponse(clientMessage);
+        if (response.error !== null) {
+            res.onExecuteError(
+                new HazelcastSqlException(
+                    response.error.originatingMemberId, response.error.code, response.error.message
+                )
+            );
+        } else {
+            res.onExecuteResponse(
+                response.rowMetadata !== null ? new SqlRowMetadataImpl(response.rowMetadata) : null,
+                response.rowPage,
+                response.updateCount
+            );
+        }
+    }
+
+    /**
+     * Validates SqlStatement
+     *
+     * @param sqlStatement
+     * @throws RangeError if validation is not successful
+     * @internal
+     */
+    static validateSqlStatement(sqlStatement: SqlStatement | null): void {
+        if (sqlStatement === null) {
+            throw new RangeError('Sql statement cannot be null');
+        }
+        tryGetString(sqlStatement.sql);
+        if (sqlStatement.hasOwnProperty('params')) // if params is provided, validate it
+            tryGetArray(sqlStatement.params);
+        if (sqlStatement.hasOwnProperty('options')) // if options is provided, validate it
+            SqlServiceImpl.validateSqlStatementOptions(sqlStatement.options);
+    }
+
+    /**
+     * Validates SqlStatementOptions
+     *
+     * @param sqlStatementOptions
+     * @throws RangeError if validation is not successful
+     * @internal
+     */
+
+    static validateSqlStatementOptions(sqlStatementOptions: SqlStatementOptions): void {
+        if (sqlStatementOptions.hasOwnProperty('schema'))
+            tryGetString(sqlStatementOptions.schema);
+
+        if (sqlStatementOptions.hasOwnProperty('timeoutMillis')) {
+            const longValue = tryGetLong(sqlStatementOptions.timeoutMillis);
+            if (longValue.lessThanOrEqual(Long.fromInt(-2))) {
+                throw new RangeError('Timeout millis cannot be less than -1');
+            }
+        }
+
+        if (sqlStatementOptions.hasOwnProperty('cursorBufferSize') && tryGetNumber(sqlStatementOptions.cursorBufferSize) <= 0) {
+            throw new RangeError('Cursor buffer size cannot be negative');
+        }
+
+        if (sqlStatementOptions.hasOwnProperty('expectedResultType'))
+            tryGetEnum(SqlExpectedResultType, sqlStatementOptions.expectedResultType);
+
+        if (sqlStatementOptions.hasOwnProperty('returnRawResult'))
+            tryGetBoolean(sqlStatementOptions.returnRawResult);
+    }
+
+    /**
+     * Converts datetime related wrapper classes to string to be able to serialize them. (No default serializers yet)
+     * @internal
+     * @param value
+     */
+    convertToStringIfDatetimeValue(value: any) {
```

Comment:
I couldn't remember exactly, but didn't we say that we don't perform any implicit conversions for date values? We suppose to write a documentation about them since those implicit casts might not work perfectly

Reply:
> We suppose to write a documentation about them since those implicit casts might not work perfectly
What do you mean? We are sending them as strings internally. This method is used to convert the wrapper classes to string.

Reply:
Implicit cast issue was writing cast to sql string instead of user writing it

Reply:
Why do we send them as strings? I remember we said we won't support sending date types as parameters in 4.2. I remember saying that, if users want to send it, they can convert it to string, add cast to their queries. We should just document this behavior, and fix it 5.0

Reply:
I don't think we need to do that. Users will send wrapper classes as parameters in 4.2, they will be converted to string internally and sent. After default serializers is added, we will send them as byte array encoded, not strings. This way the user API will be the same for both cases, and there will be no inconvenience.

Reply:
We will only make in easier by removing the need for explicit casting when the default serializers for date and time classes are added. I think sometimes doing things implicitly can help, and this is one of them

Reply:
I think we can discuss this with the team. I haven't looked at the documentation PR yet, but If we decide to send strings for these types, we should document this behavior

Reply:
I guess you are not proposing removing date/time support altogether. Users should be able to use them still. Instead of them giving strings and making it harder for them, wrapper classes are written. Wrapper classes can be uses instead of them giving strings. We can document this internal behaviour and when serializer is added they will be already familiar with the wrapper class, which is better than introducing the wrapper class later on.

Reply:
Yes, I think there still should be wrapper classes. I thought we decided to let users do things like `execute("SELECT * FROM map WHERE date < CAST(?  AS DATE)", myHzLocalDate.toString())` but not `execute("SELECT * FROM map WHERE date < CAST(?  AS DATE)", myHzLocalDate)` to make it more explicit

Reply:
We can do that if you want. No worries. Going to delete implicit string conversion

Reply:
Don't to immediately :smile: I thought we can discuss this with the team in the standup

Reply:
Ok easy to revert that commit, no worries. Should we ask it in the slack channel?

Reply:
This might be the first question to ask in tomorrow's standup